### PR TITLE
Pin Python version to 3.11

### PR DESCRIPTION
**Pull request by jiridanek**
### Don't we pin Python already?

We don't, or not strictly enough. Currently we have

https://github.com/red-hat-data-services/ods-ci/blob/afa1de64ba5e632e794d70eebb4197956270e419/pyproject.toml#L9-L10

which means "use any Python `>=3.8.1, <4.0.0`".

### Why pin Python version?

Because we _can_ put everybody on the same version.
This way,
we make it less likely different people will hit different infra issues.

### Why use Python 3.11?

It is well established,
binary wheels exist for it in PyPI,
it is available on UBI 8 and 9, and on Fedora starting with Fedora 37.

### Why say `python = ">=3.11.5, <3.12.0"`?

The syntax uses basic math instead of tildes and carets,
which not everybody might've learned what they mean.
That makes the requirement easier to read and understand.

### Will this break CI infra?

~~Most likely, which is why this PR exists, to try things out.~~

All things in order, everything is very good, 🌈 🦄 🌇

### CODE REVIEW

**Comment by jiridanek** [comment]
And, it fails CI as expected, rhods-ci-pr-test/2308

```
21:19:50  The currently activated Python version 3.9.17 is not supported by the project (>=3.11.5, <3.12.0).
21:19:50  Trying to find and use a compatible version. 
21:19:50  
21:19:50  Poetry was unable to find a compatible version. If you have one, you can explicitly use it via the "env use" command.
21:19:51  ods_ci/run_robot_test.sh: line 320: /bin/activate: No such file or directory
```

So, I'll have to update Python in CI. Then, I'll announce that contributors should do `dnf install python3.11; poetry env use /usr/bin/python3.11`, and after some time it should be safe to merge this change.

**Comment by jiridanek** [comment]
Here is a CI image update that needs to be merged first, red-hat-data-services/ods-ci/pull/1118

**Comment by jiridanek** [comment]
Also need Python 3.11 in the Dockerfile here, https://github.com/red-hat-data-services/ods-ci/blob/master/ods_ci/build/Dockerfile

File: ods_ci/run_robot_test.sh
Comment (jiridanek) [inline ods_ci/run_robot_test.sh:317]: this is not good on macos

Comment (jstourac) [reply ods_ci/run_robot_test.sh:317]: we should stop to support these :smile:

**Comment by jiridanek** [comment]
CI: rhods-ci-pr-test/2323 (smokes do run)

**Comment by jgarciao** [comment]
I've added the "do not merge" label until we have an agreement on slack about merging this

**Comment by jiridanek** [comment]
This should be all the needed changes. Now onto rechecking it in CI.

File: ods_ci/run_robot_test.sh
Comment (jiridanek) [reply ods_ci/run_robot_test.sh]: we can't, redhatters like their macs

**Comment by jiridanek** [comment]
There are two todo's that I'll do when this is merged
* [ ] go to gitlab.cee and in ods/jenkins change the pr job so that it does not archive .venv directory
* [ ] go to gitlab.cee and in ods/rhods-qe-tools make python3.11 be the default python for poetry to save CI time

File: ods_ci/run_robot_test.sh
Comment (jiridanek) [reply ods_ci/run_robot_test.sh]: done mac-friendly solution

File: ods_ci/utils/scripts/Sender/EmailSender.py
Comment (bdattoma) [inline ods_ci/utils/scripts/Sender/EmailSender.py:27]: why?

Comment (jiridanek) [reply ods_ci/utils/scripts/Sender/EmailSender.py:27]: linter wants this when you have a py3.11 project

at some point they made list[...] be an alias for typing.List[...], so that's a cleanup it wants

not sure about why it dislikes Optional, that should be fine imo, and | None is not a clear win in my eyes

**Review by bdattoma — APPROVED** [review]

**Review by jstourac — APPROVED** [review]

**Comment by jiridanek** [comment]
job/rhods/job/rhods-ci-pr-test/2487, still looks good

**Pull request merged**
